diff --git a/src/UnsavedIndexConfiguration.php b/src/UnsavedIndexConfiguration.php
index 8ec5b0bd..f1674176 100644
--- a/src/UnsavedIndexConfiguration.php
+++ b/src/UnsavedIndexConfiguration.php
@@ -17,6 +17,7 @@ use Drupal\search_api\Query\QueryInterface;
 use Drupal\search_api\Query\ResultSetInterface;
 use Drupal\search_api\Tracker\TrackerInterface;
 use Drupal\Core\TempStore\SharedTempStore;
+use Drupal\Core\Entity\EntityInterface;
 
 /**
  * Represents a configuration of an index that was not yet permanently saved.
@@ -75,6 +76,33 @@ class UnsavedIndexConfiguration implements IndexInterface, UnsavedConfigurationI
     $this->currentUserId = $current_user_id;
   }
 
+  /**
+   * Changes the values of an entity before it is duplicated.
+   *
+   * @param \Drupal\Core\Entity\EntityStorageInterface $storage
+   *   The entity storage object.
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity to be duplicated.
+   */
+  public static function preDuplicate(EntityStorageInterface $storage, EntityInterface $entity) {
+  }
+
+  /**
+   * Acts on a duplicated entity before hooks are invoked.
+   *
+   * Used after the entity is duplicated, but before saving the entity and before
+   * any of the presave hooks are invoked.
+   *
+   * See the @link entity_crud Entity CRUD topic @endlink for more information.
+   *
+   * @param \Drupal\Core\Entity\EntityStorageInterface $storage
+   *   The entity storage object.
+   *
+   * @see \Drupal\Core\Entity\EntityInterface::createDuplicate()
+   */
+  public function postDuplicate(EntityStorageInterface $storage) {
+  }
+
   /**
    * Retrieves the entity type manager.
    *
